src/main/java/io/antmedia/datastore/db/MongoStore.java

```diff
@@ -1429,4 +1432,103 @@ public boolean updateStreamMetaData(String streamId, String metaData) {
 		}
 		return false;
 	}
+
+	@Override
+	public boolean blackListToken(Token token) {
+		boolean result = false;
+		//update if exists, else insert
+		synchronized (this) {
+			if (token.getStreamId() != null && token.getTokenId() != null) {
+				Query<Token> query = tokenDatastore.find(Token.class).filter(Filters.eq(TOKEN_ID, token.getTokenId()));
+
+				if(query.first() != null){
+					final UpdateResult results = query.update(new UpdateOptions().multi(false), set(BLACKLISTED, true));
+					if(results.getModifiedCount() == 1){
+						result = true;
+					}
+				}else{
+					token.setBlackListed(true);
+					result = saveToken(token);
+				}
+			}
+		}
+
+		return result;
+
+	}
+
+	@Override
+	public Token getBlackListedToken(String tokenId) {
+		synchronized (this){
+			Query<Token> query = tokenDatastore.find(Token.class).filter(Filters.eq(TOKEN_ID, tokenId));
+			Token fetchedToken = query.first();
+			if(fetchedToken != null && fetchedToken.isBlackListed()){
+				return fetchedToken;
+			}
+		}
+		return null;
+	}
+
+	@Override
+	public boolean whiteListToken(String tokenId) {
+		synchronized (this){
+			Query<Token> query = tokenDatastore.find(Token.class).filter(Filters.eq(TOKEN_ID, tokenId));
+			final UpdateResult results = query.update(new UpdateOptions().multi(false), set(BLACKLISTED, false));
+			return results.wasAcknowledged();
+		}
+	}
+
+	@Override
+	public List<String> getBlackListedTokens() {
+		List<String> tokenBlacklist = new ArrayList<>();
+		synchronized (this){
+			Query<Token> query = tokenDatastore.find(Token.class).filter(Filters.eq(BLACKLISTED, true));
+			for (Token token : query) {
+				tokenBlacklist.add(token.getTokenId());
```

Comment:
Is it necessary to hold the monitor as long as the iteration processes all the tokens?